tests/python/pants_test/task_test_base.py

```diff
@@ -299,3 +303,116 @@ def assert_console_raises(self, exception, **kwargs):
     """
     with self.assertRaises(exception):
       self.execute_console_task(**kwargs)
+
+
+class DeclarativeTaskTestMixin(object):
+  """Experimental mixin combining target descriptions with a file path dict.
+
+  This class should be mixed in to subclasses of `TaskTestBase`!
+
+  NB: `self.populate_target_dict()` should be called in the `setUp()` method to use the target specs
+  specified in `dist_specs`!
+
+  This mixin also allows specifying tasks to be run before or after the task_type() is executed when
+  calling `self.invoke_tasks()`.
+  """
+
+  @classproperty
+  def dist_specs(cls):
+    """This is an informally-specified nested dict.
+
+    Special keys are 'key' (used to index into `self.target_dict`) and 'filemap' (creates files at
+    the specified relative paths). The rest of the keys are fed into `self.make_target()`. An
+    `OrderedDict` of 2-tuples may be used if targets need to be created in a specific order (e.g. if
+    they have dependencies on each other).
+    """
+    raise NotImplementedError('dist_specs must be implemented!')
+
+  @classproperty
+  def run_before_task_types(cls):
```

Comment:
The `run_before` and `run_after` tasks look very important for correct usage of this class, but it's not clear to me what they would do (without reading the implementation of this class).

Reply:
Added lots more documentation (and removed the previous docstrings)!